codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -692,9 +671,185 @@ void decrementFlowControlWindow(int bytes) {
          * Discards this {@link FlowControlled}, writing an error. If this frame is in the pending queue,
          * the unwritten bytes are removed from this branch of the priority tree.
          */
-        void writeError(FlowControlled frame, Http2Exception cause) {
+        private void writeError(FlowControlled frame, Http2Exception cause) {
             decrementPendingBytes(frame.size());
             frame.error(cause);
         }
     }
+
+    /**
+     * The remote flow control state for a single stream that is not in a state where flow controlled frames cannot
+     * be exchanged.
+     */
+    private static final class ReducedFlowState extends RemoteFlowControlState {
+        public ReducedFlowState(Http2Stream stream) {
+            super(stream);
+        }
+
+        public ReducedFlowState(RemoteFlowControlState existingState) {
+            super(existingState);
+        }
+
+        @Override
+        public int windowSize() {
+            return 0;
+        }
+
+        @Override
+        public int initialWindowSize() {
+            return 0;
+        }
+
+        @Override
+        public int writableWindow() {
+            return 0;
+        }
+
+        @Override
+        public int streamableBytes() {
+            return 0;
+        }
+
+        @Override
+        public int streamableBytesForTree() {
+            return streamableBytesForTree;
+        }
+
+        @Override
+        public void writeAllocatedBytes() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void cancel() {
+        }
+
+        @Override
+        public void window(int initialWindowSize) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public int incrementStreamWindow(int delta) throws Http2Exception {
+            // This operation needs to be supported during the initial settings exchange when
+            // the peer has not yet acknowledged this peer being activated.
+            return 0;
+        }
+
+        @Override
+        public int writeBytes(int bytes) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void enqueueFrame(FlowControlled frame) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void allocate(int bytes) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean hasFrame() {
+            return false;
+        }
+    }
+
+    /**
+     * An abstraction around {@link FlowControlState} which provides specific extensions used by remote flow control.
+     */
+    private abstract static class RemoteFlowControlState implements FlowControlState {
```

Comment:
I would like to keep consistent with the naming consistent with the `Local` side too (so it is easier to see the conceptual relationship), but for the `Local` side we have an interface.  Using `Base` for an interfaces seems a bit awkward.  I also think `RemoteFlowControlState` and `LocalFlowControlState` is a natural extension to `FlowControlState`.  I know we are already in the context of `Remote` or `Local` because we are in the flow controller but I'm not sure if changing these names clarifies the situation.  WDTY?